packages/flutter_tools/test/general.shard/windows/visual_studio_test.dart

```diff
@@ -3,216 +3,246 @@
-        '-latest',
-        ...?additionalArguments,
-        ...?requirementArguments,
-      ],
-      workingDirectory: anyNamed('workingDirectory'),
-      environment: anyNamed('environment'),
-    )).thenAnswer((Invocation invocation) {
-      return result;
-    });
-  }
+const String programFilesPath = r'C:\Program Files (x86)';
+const String visualStudioPath = programFilesPath + r'\Microsoft Visual Studio\2017\Community';
+const String vcvarsPath = visualStudioPath + r'\VC\Auxiliary\Build\vcvars64.bat';
+const String vswherePath = programFilesPath + r'\Microsoft Visual Studio\Installer\vswhere.exe';
 
-  // Sets whether or not a vswhere query with the required components will
-  // return an installation.
-  void setMockCompatibleVisualStudioInstallation(Map<String, dynamic>response) {
-    setMockVswhereResponse(_requiredComponents, <String>['-version', '16'], response);
+final Platform windowsPlatform = FakePlatform(
+  operatingSystem: 'windows',
+  environment: <String, String>{
+    'PROGRAMFILES(X86)': r'C:\Program Files (x86)\',
   }
+);
+
+// A minimum version of a response where a VS installation was found.
+const Map<String, dynamic> _defaultResponse = <String, dynamic>{
+  'installationPath': visualStudioPath,
+  'displayName': 'Visual Studio Community 2019',
+  'installationVersion': '16.2.29306.81',
+  'isRebootRequired': false,
+  'isComplete': true,
+  'isLaunchable': true,
+  'isPrerelease': false,
+  'catalog': <String, dynamic>{
+    'productDisplayVersion': '16.2.5',
+  },
+};
+
+// A response for a VS installation that's too old.
+const Map<String, dynamic> _tooOldResponse = <String, dynamic>{
+  'installationPath': visualStudioPath,
+  'displayName': 'Visual Studio Community 2017',
+  'installationVersion': '15.9.28307.665',
+  'isRebootRequired': false,
+  'isComplete': true,
+  'isLaunchable': true,
+  'isPrerelease': false,
+  'catalog': <String, dynamic>{
+    'productDisplayVersion': '15.9.12',
+  },
+};
+
+// A version of a response that doesn't include certain installation status
+// information that might be missing in older vswhere.
+const Map<String, dynamic> _missingStatusResponse = <String, dynamic>{
+  'installationPath': visualStudioPath,
+  'displayName': 'Visual Studio Community 2017',
+  'installationVersion': '16.4.29609.76',
+  'catalog': <String, dynamic>{
+    'productDisplayVersion': '16.4.1',
+  },
+};
+
+// Arguments for a vswhere query to search for an installation with the required components.
+const List<String> _requiredComponents = <String>[
+  'Microsoft.Component.MSBuild',
+  'Microsoft.VisualStudio.Component.VC.Tools.x86.x64',
+  'Microsoft.VisualStudio.Component.Windows10SDK.17763',
+];
+
+
+// Sets up the mock environment so that searching for Visual Studio with
+// exactly the given required components will provide a result. By default it
+// return a preset installation, but the response can be overridden.
+void setMockVswhereResponse(
+  VisualStudio visualStudio, [
```

Comment:
This is where the object-under-test-as-fixture is showing up (per other comment); a vswhere response shouldn't involve a VisualStudio object at all.

I think tests are much clearer if there's a separation between setting up the environment the object is being tested in, and then making and testing the object in that environment. Short term, that could mean this would operate on fixture-level FileSystem and ProcessManager objects as is currently common in the tool tests. Longer term (or shorter, if you want to do it now, but I don't think it's as important as the larger change), a couple of ways this could be done:
- Make a dumb data object in this file that holds those two objects; tests would then call a function to configure the environment, which would return that data object, then they would create the VS object by passing in the things from them. In other words, use a local bag of objects instead of an enclosing fixture.
- Do dependency injection for test helpers the same way you are now doing for real objects. So a test would look something like:
  ```
  FileSystem fileSystem = ...
  ProcessManager processManager = ...
  setMockVswhereResponse(fileSystem: fileSystem, processManager: processManager, ...);
  
  VisualStudio visualStudio = VisualStudio(
    fileSystem: fileSystem,
    processManager: processManager,
    ...
  );
  ```

Reply:
I sort of did a mix of the two: I created a simple test fixture so I could have a setUp method that returns everything I need. Then, I'm drilling the FileSystem and ProcessManager down into the test helpers